presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java

```diff
@@ -86,24 +90,40 @@
     private final Map<Integer, Type> columnTypes;                     // key: index into hiveColumnIndices array
     private final Object[] constantValues;                            // aligned with hiveColumnIndices array
     private final Function<Block, Block>[] coercers;                   // aligned with hiveColumnIndices array
-    private final List<FilterFunction> filterFunctionsWithInputs;
+
+    // non-deterministic filter function with no inputs (rand() < 0.1); evaluated before any column is read
     private final Optional<FilterFunction> filterFunctionWithoutInput;
     private final Map<Integer, Integer> filterFunctionInputMapping;   // channel-to-index-into-hiveColumnIndices-array mapping
-    private final Set<Integer> filterFunctionInputs;                  // channels
     private final Map<Integer, Integer> columnsWithFilterScores;      // keys are indices into hiveColumnIndices array; values are filter scores
 
     // Optimal order of stream readers
     private int[] streamReaderOrder;                                  // elements are indices into hiveColumnIndices array
 
+    private List<FilterFunction>[] filterFunctionsOrder;              // aligned with streamReaderOrder order; each filter function is placed
+                                                                      // into a list positioned at the last necessary input
+    private Set<Integer>[] filterFunctionInputs;                      // aligned with filterFunctionsOrder
+
+    // non-deterministic filter functions with only constant inputs; evaluated before any column is read
+    private List<FilterFunction> filterFunctionsWithConstantInputs;
+    private Set<Integer> filterFunctionConstantInputs;
+
     // An immutable list of initial positions; includes all positions: 0,1,2,3,4,..
     // This array may grow, but cannot shrink. The values don't change.
     private int[] positions;
 
     // Used in applyFilterFunctions; mutable
     private int[] outputPositions;
     private RuntimeException[] errors;
+    private RuntimeException[] tmpErrors;
```

Comment:
How about `intermediateErrors`? Maybe add a comment to explain the difference from the above

Reply:
@highker The difference is that `errors` contains shared state (used across method calls), but `tmpErrors` is just a local variable (I'm using member variable only to enable memory re-use). How about this?

```
    // errors encountered while evaluating filter functions; indices are positions in the batch
    // of rows being processed by getNextPage (errors[outputPositions[i]] is valid)
    private RuntimeException[] errors;

    // temporary array to be used in applyFilterFunctions only; exists solely for the purpose of re-using memory
    // indices are positions in a page provided to the filter filters (it contains a subset of rows that passed earlier filters)
    private RuntimeException[] tmpErrors;
```